### PR TITLE
[8 0f #107] move base env tool providers to dedicated files

### PR DESCRIPTION
**Pull request by bollwyvl**
This moves extra tools for the `base` env into dedicated files, and finally reduces `setup.ts` to _right at_ 100 lines.

This is the end of what #107 had implemented, and no additional changes are planned at this scale :tada: 

- closes #107

Thanks @goanpeca for the review

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -4,17 +4,26 @@ import * as stream from "stream";
 
 import * as exec from "@actions/exec";
 import * as core from "@actions/core";
+import * as constants from "./constants";
 
 import {
   CONDA_CACHE_FOLDER,
   FORCED_ERRORS,
   IGNORED_WARNINGS,
 } from "./constants";
```
Comment (goanpeca) [inline src/utils.ts:13]: Wondering why we import direct `consants` but also use `constants.SOMETHING`, should we just use `constants.SOMETHING` for everything?

File: src/types.ts [hunk 2]
```diff
@@ -153,3 +153,37 @@ export interface IEnvProvider {
...
+  label: string;
+  /**
+   * Whether this provider is requested by action inputs
+   */
+  provides: (
+    inputs: IActionInputs,
+    options: IDynamicOptions
+  ) => Promise<boolean>;
+  /**
+   * Conda package specs and option yupdates for tools to install after updating
```
Comment (goanpeca) [inline src/types.ts:176]: yupdates - updates

File: src/types.ts [hunk 3]
```diff
@@ -153,3 +153,37 @@ export interface IEnvProvider {
...
+   */
+  toolPackages: (
+    inputs: IActionInputs,
+    options: IDynamicOptions
+  ) => Promise<IToolUpdates>;
+  /**
+   * Steps to peform after the env is updated, and potentially reconfigured
```
Comment (goanpeca) [inline src/types.ts:183]: peform -> perform

File: src/types.ts [hunk 4]
```diff
@@ -153,3 +153,37 @@ export interface IEnvProvider {
...
+   */
+  postInstall?: (
```
Comment (goanpeca) [inline src/types.ts:185]: Cute :)

File: src/base-tools/index.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import { updateCondaBuild } from "./update-conda-build";
+
+/**
+ * The providers of tool updates: order isn't _really_ important
+ *
+ * ### Note
+ * To add a new tool provider,
+ * - implement IToolProvider and add it here
+ * - probably add inputs to `../../action.yaml`
+ * - any any new RULEs in ../input.ts, for example if certain inputs make no sense
```
Comment (goanpeca) [inline src/base-tools/index.ts:19]: add any ?

**Review by goanpeca — APPROVED** [review]
Thanks @bollwyvl for the awesome refactoring work!

LGTM :)

File: src/utils.ts (hunk 1 above)
Comment (bollwyvl) [reply src/utils.ts:13]: yes, will fix

File: src/base-tools/index.ts (hunk 5 above)
Comment (bollwyvl) [reply src/base-tools/index.ts]: fixed in 8ed133d

File: src/types.ts (hunk 4 above)
Comment (bollwyvl) [reply src/types.ts:185]: ?

File: src/types.ts (hunk 3 above)
Comment (bollwyvl) [reply src/types.ts]: fixed in 8ed133d

File: src/utils.ts (hunk 1 above)
Comment (bollwyvl) [reply src/utils.ts]: fixed in 8ed133d

**Comment by bollwyvl** [comment]
@goanpeca thanks for the quick review... think i hit all the points raised above...

**Comment by bollwyvl** [comment]
Not _quite_ready... 
```

Thu, 24 Dec 2020 18:30:26 GMT   
Thu, 24 Dec 2020 18:30:26 GMT   Currently, only install, create, list, search, run, info and clean are supported through mamba.
Thu, 24 Dec 2020 18:30:26 GMT   Creating bash wrapper for `mamba`...
Thu, 24 Dec 2020 18:30:26 GMT   ... wrote C:/Miniconda\condabin\mamba.bat:
Thu, 24 Dec 2020 18:30:26 GMT   bash.exe -c "exec 'C:/Miniconda\condabin\mamba.bat' $*"
```

I want those slashes to be consistent...

**Comment by bollwyvl** [comment]
Looks better:
```
   Creating bash wrapper for `mamba`...
  ... wrote C:/Miniconda/condabin/mamba.bat:
  bash.exe -c "exec 'C:/Miniconda/condabin/mamba.bat' $*" || exit 1
```

**Comment by bollwyvl** [comment]
Ok, I think the log output is good now... added a final output, just so there's a timestamp outside the group.

File: src/types.ts (hunk 4 above)
Comment (goanpeca) [reply src/types.ts:185]: Haha nothing :)

**Pull request merged**